### PR TITLE
Add predict method for reconstruction

### PR DESCRIPTION
**Pull request by ravih18**
Hello @clementchadebec ,

As discussed on Thursday, here is a small PR with two contribution:
* homogenizing the forward method outputs for all the models, the key for the reconstruction loss is `recon_loss`. If you think `reconstruction_loss` is better I can change it for the second option.
* Adding a `predict` method to the BaseAE class in order to compute the model outputs (reconstruction and embedding) without having to compute the loss or posterior parameters (embedding is equal to z in inference mode if I understand well). This might be overlapping with the `reconstruct`, let me know what di you think of it, it might not be useful to add to the main code.

Have a good week, 
Ravi

### CODE REVIEW

**Comment by soumickmj** [comment]
Nice idea!
I was also thinking to implement a similar method for my task.
Thanks :)

File: src/pythae/models/beta_tc_vae/beta_tc_vae_model.py [hunk 1]
```diff
@@ -79,7 +79,7 @@ def forward(self, inputs: BaseDataset, **kwargs):
         )
 
         output = ModelOutput(
-            reconstruction_loss=recon_loss,
```
Comment (clementchadebec) [inline src/pythae/models/beta_tc_vae/beta_tc_vae_model.py:82]: Update this [test](https://github.com/ravih18/pythae/blob/a1c8cce2be8e94ecb7283e7f65e568581b82df14/tests/test_BetaTCVAE.py#L295-L297) in `test_BetaTCVAE` with `recon_loss` to avoid failure

File: src/pythae/models/beta_vae/beta_vae_model.py [hunk 2]
```diff
@@ -71,7 +71,7 @@ def forward(self, inputs: BaseDataset, **kwargs):
         loss, recon_loss, kld = self.loss_function(recon_x, x, mu, log_var, z)
 
         output = ModelOutput(
-            reconstruction_loss=recon_loss,
```
Comment (clementchadebec) [inline src/pythae/models/beta_vae/beta_vae_model.py:74]: Update this [test](https://github.com/ravih18/pythae/blob/a1c8cce2be8e94ecb7283e7f65e568581b82df14/tests/test_BetaVAE.py#L287-L289) in `test_BetaVAE` with `recon_loss` to avoid failure

File: src/pythae/models/ciwae/ciwae_model.py [hunk 3]
```diff
@@ -78,7 +78,7 @@ def forward(self, inputs: BaseDataset, **kwargs):
         loss, recon_loss, kld = self.loss_function(recon_x, x, mu, log_var, z)
 
         output = ModelOutput(
-            reconstruction_loss=recon_loss,
```
Comment (clementchadebec) [inline src/pythae/models/ciwae/ciwae_model.py:81]: Update test as well to avoid failure

File: src/pythae/models/disentangled_beta_vae/disentangled_beta_vae_model.py [hunk 4]
```diff
@@ -79,7 +79,7 @@ def forward(self, inputs: BaseDataset, **kwargs):
         loss, recon_loss, kld = self.loss_function(recon_x, x, mu, log_var, z, epoch)
 
         output = ModelOutput(
-            reconstruction_loss=recon_loss,
```
Comment (clementchadebec) [inline src/pythae/models/disentangled_beta_vae/disentangled_beta_vae_model.py:82]: Update test as well to avoid failure

File: src/pythae/models/info_vae/info_vae_model.py [hunk 5]
```diff
@@ -77,7 +77,7 @@ def forward(self, inputs: BaseDataset, **kwargs) -> ModelOutput:
 
         output = ModelOutput(
             loss=loss,
-            reconstruction_loss=recon_loss,
```
Comment (clementchadebec) [inline src/pythae/models/info_vae/info_vae_model.py:80]: Update test as well to avoid failure

File: src/pythae/models/iwae/iwae_model.py [hunk 6]
```diff
@@ -77,7 +77,7 @@ def forward(self, inputs: BaseDataset, **kwargs):
         loss, recon_loss, kld = self.loss_function(recon_x, x, mu, log_var, z)
 
         output = ModelOutput(
-            reconstruction_loss=recon_loss,
```
Comment (clementchadebec) [inline src/pythae/models/iwae/iwae_model.py:80]: Update test as well to avoid failure

File: src/pythae/models/miwae/miwae_model.py [hunk 7]
```diff
@@ -86,7 +86,7 @@ def forward(self, inputs: BaseDataset, **kwargs):
         loss, recon_loss, kld = self.loss_function(recon_x, x, mu, log_var, z)
 
         output = ModelOutput(
-            reconstruction_loss=recon_loss,
```
Comment (clementchadebec) [inline src/pythae/models/miwae/miwae_model.py:89]: Update test as well to avoid failure

File: src/pythae/models/msssim_vae/msssim_vae_model.py [hunk 8]
```diff
@@ -72,7 +72,7 @@ def forward(self, inputs: BaseDataset, **kwargs):
         loss, recon_loss, kld = self.loss_function(recon_x, x, mu, log_var, z)
 
         output = ModelOutput(
-            reconstruction_loss=recon_loss,
```
Comment (clementchadebec) [inline src/pythae/models/msssim_vae/msssim_vae_model.py:75]: Update test as well to avoid failure

File: src/pythae/models/piwae/piwae_model.py [hunk 9]
```diff
@@ -90,7 +90,7 @@ def forward(self, inputs: BaseDataset, **kwargs):
         loss = miwae_loss + iwae_loss
 
         output = ModelOutput(
-            reconstruction_loss=recon_loss,
```
Comment (clementchadebec) [inline src/pythae/models/piwae/piwae_model.py:93]: Update test as well to avoid failure

File: src/pythae/models/pvae/pvae_model.py [hunk 10]
```diff
@@ -115,7 +115,7 @@ def forward(self, inputs: BaseDataset, **kwargs):
         loss, recon_loss, kld = self.loss_function(recon_x, x, z, qz_x)
 
         output = ModelOutput(
-            reconstruction_loss=recon_loss,
```
Comment (clementchadebec) [inline src/pythae/models/pvae/pvae_model.py:118]: Update test as well to avoid failure

File: src/pythae/models/svae/svae_model.py [hunk 11]
```diff
@@ -88,7 +88,7 @@ def forward(self, inputs: BaseDataset, **kwargs):
         loss, recon_loss, kld = self.loss_function(recon_x, x, loc, concentration, z)
 
         output = ModelOutput(
-            reconstruction_loss=recon_loss,
```
Comment (clementchadebec) [inline src/pythae/models/svae/svae_model.py:91]: Update test as well to avoid failure

File: src/pythae/models/vae/vae_model.py [hunk 12]
```diff
@@ -88,7 +88,7 @@ def forward(self, inputs: BaseDataset, **kwargs):
         loss, recon_loss, kld = self.loss_function(recon_x, x, mu, log_var, z)
 
         output = ModelOutput(
-            reconstruction_loss=recon_loss,
```
Comment (clementchadebec) [inline src/pythae/models/vae/vae_model.py:91]: Update test as well to avoid failure

File: src/pythae/models/vae_iaf/vae_iaf_model.py [hunk 13]
```diff
@@ -93,7 +93,7 @@ def forward(self, inputs: BaseDataset, **kwargs):
         )
 
         output = ModelOutput(
-            reconstruction_loss=recon_loss,
```
Comment (clementchadebec) [inline src/pythae/models/vae_iaf/vae_iaf_model.py:96]: Update test as well to avoid failure

File: src/pythae/models/vae_lin_nf/vae_lin_nf_model.py [hunk 14]
```diff
@@ -106,7 +106,7 @@ def forward(self, inputs: BaseDataset, **kwargs):
         )
 
         output = ModelOutput(
-            reconstruction_loss=recon_loss,
```
Comment (clementchadebec) [inline src/pythae/models/vae_lin_nf/vae_lin_nf_model.py:109]: Update test as well to avoid failure

File: src/pythae/models/vamp/vamp_model.py [hunk 15]
```diff
@@ -91,7 +91,7 @@ def forward(self, inputs: BaseDataset, **kwargs):
         loss, recon_loss, kld = self.loss_function(recon_x, x, mu, log_var, z, epoch)
 
         output = ModelOutput(
-            reconstruction_loss=recon_loss,
```
Comment (clementchadebec) [inline src/pythae/models/vamp/vamp_model.py:94]: Update test as well to avoid failure

**Review by clementchadebec — CHANGES_REQUESTED** [review]
Hi @ravih18, thanks again for this nice contribution. Everything looks good but two small things:
- I think we should add a test for the `predict` method you proposed (see comments)
- We should update the tests `test_model_train_output` with the proper `recon_loss` key for all the models where the key changed (see comments).

Thanks again!

**Comment by clementchadebec** [comment]
Hello @ravih18, thank you again for this nice contribution!
I agree that adding a `predict` method is indeed a good idea especially for downstream tasks that do not require computing the loss. It now allows to easily recover the reconstruction **and** embeddings of any model.

I just had a look to the PR and have put some comments in the review :)

**Comment by clementchadebec** [comment]
Closing since solved in #75
